modules/solid_mechanics/src/materials/EigenDecompMaterial.C

```diff
@@ -0,0 +1,65 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "EigenDecompMaterial.h"
+#include "RankTwoTensor.h"
+
+registerMooseObject("SolidMechanicsApp", EigenDecompMaterial);
+registerMooseObject("SolidMechanicsApp", ADEigenDecompMaterial);
+
+template <bool is_ad>
+InputParameters
+EigenDecompMaterialTempl<is_ad>::validParams()
+{
+  InputParameters params = Material::validParams();
+  params.addClassDescription(
+      "Creates material with eigenvalues and eigenvectors of a symmetric rank two tensor.");
+  params.addRequiredParam<MaterialPropertyName>(
+      "rank_two_tensor",
+      "The name of the symmetric rank two tensor to used in eigen decomposition.");
+  params.addParam<std::string>(
+      "base_name",
+      "Optional parameter to allow multiple tensors to be decomposed on the same block.");
+  return params;
+}
+
+template <bool is_ad>
+EigenDecompMaterialTempl<is_ad>::EigenDecompMaterialTempl(const InputParameters & parameters)
+  : Material(parameters),
+    _base_name(isParamValid("base_name") ? getParam<std::string>("base_name") + "_" : ""),
+    _tensor(getGenericMaterialProperty<RankTwoTensor, is_ad>("rank_two_tensor")),
+    _max_eigen_vector(
+        declareGenericProperty<RealVectorValue, is_ad>(_base_name + "max_eigen_vector")),
+    _mid_eigen_vector(
+        declareGenericProperty<RealVectorValue, is_ad>(_base_name + "mid_eigen_vector")),
+    _min_eigen_vector(
+        declareGenericProperty<RealVectorValue, is_ad>(_base_name + "min_eigen_vector")),
+    _max_eigen_value(declareGenericProperty<Real, is_ad>(_base_name + "max_eigen_value")),
+    _mid_eigen_value(declareGenericProperty<Real, is_ad>(_base_name + "mid_eigen_value")),
+    _min_eigen_value(declareGenericProperty<Real, is_ad>(_base_name + "min_eigen_value"))
```

Comment:
This naming scheme assumes that `LIBMESH_DIM` is always 3. Realistically that's probably required in MOOSE, but yet we do have over 1600 checks for `LIBMESH_DIM` in our code. I suggest checking the dimension.